x/token/keeper/transfer.go

```diff
@@ -0,0 +1,193 @@
+package keeper
+
+import (
+	"github.com/gogo/protobuf/proto"
+	sdk "github.com/line/lbm-sdk/types"
+	sdkerrors "github.com/line/lbm-sdk/types/errors"
+	"github.com/line/lbm-sdk/x/token"
+)
+
+func (k Keeper) Transfer(ctx sdk.Context, from, to sdk.AccAddress, amounts []token.FT) error {
+	if err := k.transfer(ctx, from, to, amounts); err != nil {
+		return err
+	}
+
+	events := make([]proto.Message, 0, len(amounts))
+	for _, amount := range amounts {
+		events = append(events, &token.EventTransfer{
+			ClassId: amount.ClassId,
+			From:    from.String(),
+			To:      to.String(),
+			Amount:  amount.Amount,
+		})
+	}
+	return ctx.EventManager().EmitTypedEvents(events...)
+}
+
+func (k Keeper) transfer(ctx sdk.Context, from, to sdk.AccAddress, amounts []token.FT) error {
+	if err := k.sendTokens(ctx, from, to, amounts); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (k Keeper) TransferFrom(ctx sdk.Context, proxy, from, to sdk.AccAddress, amounts []token.FT) error {
+	if err := k.transferFrom(ctx, proxy, from, to, amounts); err != nil {
+		return err
+	}
+
+	events := make([]proto.Message, 0, len(amounts))
+	for _, amount := range amounts {
+		events = append(events, &token.EventTransfer{
+			ClassId: amount.ClassId,
+			From:    from.String(),
+			To:      to.String(),
+			Amount:  amount.Amount,
+		})
+	}
+	return ctx.EventManager().EmitTypedEvents(events...)
+}
+
+func (k Keeper) transferFrom(ctx sdk.Context, proxy, from, to sdk.AccAddress, amounts []token.FT) error {
+	for _, amount := range amounts {
+		if ok := k.GetApprove(ctx, from, proxy, amount.ClassId); !ok {
+			return sdkerrors.Wrapf(sdkerrors.ErrUnauthorized, "%s is not authorized to send %s tokens of %s", proxy, amount.ClassId, from)
+		}
+	}
+
+	return k.transfer(ctx, from, to, amounts)
+}
+
+func (k Keeper) Approve(ctx sdk.Context, approver, proxy sdk.AccAddress, classID string) error {
+	if err := k.approve(ctx, approver, proxy, classID); err != nil {
+		return err
+	}
+
+	return ctx.EventManager().EmitTypedEvent(&token.EventApprove{
+		ClassId:  classID,
+		Approver: approver.String(),
+		Proxy:    proxy.String(),
+	})
+}
+
+func (k Keeper) approve(ctx sdk.Context, approver, proxy sdk.AccAddress, classID string) error {
+	if k.GetApprove(ctx, approver, proxy, classID) {
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "Already approved")
+	}
+	k.setApprove(ctx, approver, proxy, classID, true)
+	return nil
+}
+
+func (k Keeper) GetApprove(ctx sdk.Context, approver, proxy sdk.AccAddress, classID string) bool {
+	store := ctx.KVStore(k.storeKey)
+	return store.Has(approveKey(classID, proxy, approver))
+}
+
+func (k Keeper) setApprove(ctx sdk.Context, approver, proxy sdk.AccAddress, classID string, set bool) {
```

Comment:
When does the approved info delete?
Is there the case to delete approved info in the lbm-sdk?

Reply:
No, there is no such a case. You cannot "disapprove" an existing approval on lbm-sdk.
The previous implementation on line/link does not have one, so I had no choice.

Reply:
OK, I understand.
@mizip, could you check this?

Reply:
Sorry for late response.
It seems you'd like to know whether "disapprove" is necessary or not in the token module right?
Based what I know, BPM team(PL1) is already aware of this and didn't request related to "disapprove" so far.
For your information, unlike token module, the collection module supports disapprove but no API is supported for that on LBD so far.
I will double check on the chain transition track and get back to you.

Reply:
Could you let me know the difference btw approve, GetApprove, SetApprove?

Reply:
> Could you let me know the difference btw approve, GetApprove, SetApprove?

Except for GetApprove, the others are private, so you don't have to worry about it.
GetApprove: get the approval information.
FYI:
approve: a private function which adds approval, but does not emit events.
setApprove: sets an approval. It's a quite low-level function and does not check constraints.